### PR TITLE
RFC: Phonebook backend

### PR DESCRIPTION
**Pull request by agger-magenta**
**NOTE:** I don't have any intention of merging this to master yet. If I had had it on a feature branch, I might (still) merge it to development. 

This is a request for comments - please criticize as you would if it were a real pull request. :-)

### CODE REVIEW

File: backend/import/do_import.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    def writer(data):
+        out_file = "{}.json".format(data[field_name].replace(' ', ''))
+        out_file = os.path.join(target_dir, out_file)
+        with open(out_file, 'w') as f:
+            json.dump(data, f)
+
+    return writer
+
+
+if __name__ == '__main__':
```
Comment (danchr) [inline backend/import/do_import.py]: Take a look at [click](http://click.palletsprojects.com) 🙂

File: backend/import/mo_api.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+DEFAULT_MO_URL = os.environ.get(
+        'MO_URL', 'http://morademo.atlas.magenta.dk/service'
+)
+ORG_ROOT = os.environ.get(
+    'MO_ORG_ROOT', '293089ba-a1d7-4fff-a9d0-79bd8bab4e5b'
+)
+
+
+def mo_get(url):
```
Comment (danchr) [inline backend/import/mo_api.py]: You should move this to an instance method on `MOData`, and stash an instance of `requests.Session` on the object itself. Sessions can provide a noticeable speedup through connection pooling and HTTP pipelining.

File: backend/import/mo_api.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+from collections import defaultdict
+
+import requests
+
+from cached_property import cached_property
+
+
+DEFAULT_MO_URL = os.environ.get(
+        'MO_URL', 'http://morademo.atlas.magenta.dk/service'
```
Comment (danchr) [inline backend/import/mo_api.py]: This should be configurable 🙂 Preferably using a JSON file somewhere for deployment.

File: backend/import/mo_api.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+)
+ORG_ROOT = os.environ.get(
+    'MO_ORG_ROOT', '293089ba-a1d7-4fff-a9d0-79bd8bab4e5b'
```
Comment (danchr) [inline backend/import/mo_api.py]: Ditto.

File: backend/import/mo_api.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+def get_ous(org_id=ORG_ROOT, mo_url=DEFAULT_MO_URL):
+    """Get all organization units belonging to org_id."""
+    return mo_get(mo_url + '/o/' + org_id + '/ou/')['items']
+
+
+def get_employees(org_id=ORG_ROOT, mo_url=DEFAULT_MO_URL):
+    return mo_get(mo_url + '/o/' + org_id + '/e/')['items']
+
+
+if __name__ == '__main__':
```
Comment (danchr) [inline backend/import/mo_api.py]: Click 🙂

File: backend/import/mo_api.py (hunk 2 above)
Comment (danchr) [reply backend/import/mo_api.py]: The implementation is [here](https://github.com/requests/requests/blob/v2.20.1/requests/api.py#L59): 

```python
def request(method, url, **kwargs):
[snip]
    # By using the 'with' statement we are sure the session is closed, thus we
    # avoid leaving sockets open which can trigger a ResourceWarning in some
    # cases, and look like a memory leak in others.
    with sessions.Session() as session:
        return session.request(method=method, url=url, **kwargs)
```

The documentation is [here](http://docs.python-requests.org/en/master/user/advanced/).

File: backend/import/mo_api.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        return mo_get(self.url + '/details/')
+
+    @cached_property
+    def detail_fields(self):
+        return list(self._details.keys())
+
+    def _get_detail(self, detail):
+        return mo_get(self.url + '/details/' + detail)
+
+    def __getattr__(self, name):
```
Comment (danchr) [inline backend/import/mo_api.py]: I'd make this `__getitem__` instead — `__getattr__` is somewhat magical.

File: backend/import/do_import.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    # Parent - UUID if exists, ROOT if not.
+    parent = ou.json['parent']['uuid'] if ou.json['parent'] else 'ROOT'
+    # For employees, we need job function, name and UUID.
+    employees = [
+        (
+            e['person']['name'], e['person']['uuid'], e['job_function']['name']
+        ) for e in ou.engagement
+    ]
+    # For associates: Association type, job function, name and UUID.
+    associates = [
```
Comment (danchr) [inline backend/import/do_import.py]: Verbage: an associate is a person, whereas what we represent as `association` is not.

File: backend/import/mo_api.py (hunk 3 above)
Comment (agger-magenta) [reply backend/import/mo_api.py]: It _is_ configurable. Just set the environment variable `MO_URL` and that gives you the value. 

Whether or not to use environment variables for this is, of course, a choice.

File: backend/import/mo_api.py (hunk 2 above)
Comment (agger-magenta) [reply backend/import/mo_api.py]: I will look into that! If it's a instance method on MOData, we'll always have one session per thread in the current implementation, which should be fine.

File: backend/import/mo_api.py (hunk 6 above)
Comment (agger-magenta) [reply backend/import/mo_api.py]: I like the magic. :-)

Anyway, I'm distinguishing between object members, which get stuff from MO and whose values are dictionaries, and data, which are accessed by ` __getitem__`. I like the idea of `MOData` as an object with members, and the distinction makes the code a bit cleaner, I think.

File: backend/import/do_import.py (hunk 7 above)
Comment (agger-magenta) [reply backend/import/do_import.py]: ... but an association seems to always contain one person in the current data set? Or is that a simplification?

Comment (agger-magenta) [reply backend/import/do_import.py]: The people referred to here are listed as "tilknyttede" in the GUI. 

The people referred to in "engament" are listed as "medarbejdere".

Comment (danchr) [reply backend/import/do_import.py]: [definition of associate on Merriam-Webster](https://www.merriam-webster.com/dictionary/associate)

Comment (danchr) [reply backend/import/do_import.py]: I'd be happy if you called it “associated” instead — “associates” just feels inaccurrate 🙂

**Comment by agger-magenta** [comment]
Tak :-)

**Comment by agger-magenta** [comment]
RFC approved, updated with comments from review.

**Pull request closed without merging**